File: examples/tools/typescript/tool.mjs
#!/usr/bin/env node
/**
 * json_query — Node.js ESM fallback (no TypeScript compilation required).
 *
 * Run with: node tool.mjs
 *
 * For the TypeScript version: deno run tool.ts
 * (or: npx tsx tool.ts)
 */

import { createInterface } from 'node:readline';

const DEFINITION = {
  name: 'json_query',
  description:
    'Query or transform JSON data using dot-notation paths. ' +
    'Useful for extracting values from API responses, config files, or structured data. ' +
    'Supports dot-path extraction, key listing, array indexing, and pretty-printing.',
  parameters: {
    type: 'object',
    properties: {
      json: {
        type: 'string',
        description: 'JSON string to query. Must be valid JSON.',
      },
      path: {
        type: 'string',
        description:
          "Dot-notation path to extract, e.g. 'user.address.city' or 'items[0].name'. " +
          "Use '.' for the root object. Use '*' to list all keys.",
      },
      transform: {
        type: 'string',
        description:
          "Optional transformation: 'keys', 'values', 'length', 'type', 'pretty', 'compact'.",
      },
    },
    required: ['json', 'path'],
  },
};

function getPath(obj, path) {
  if (path === '.' || path === '') return obj;

  const parts = [];
  for (const segment of path.split('.')) {
    if (segment === '') continue; // leading/trailing dot
    const arrMatch = segment.match(/^(.*?)\[(\d+)\]$/);
    if (arrMatch) {
      if (arrMatch[1]) parts.push(arrMatch[1]);
      parts.push(parseInt(arrMatch[2], 10));
    } else {
      parts.push(segment);
    }
  }

  let current = obj;
  for (const part of parts) {
    if (current === null || current === undefined) {
      throw new Error(`Path segment "${part}" not found — parent is null`);
    }
    if (typeof part === 'number') {
      if (!Array.isArray(current)) throw new Error(`Cannot index non-array with [${part}]`);
      current = current[part];
    } else if (part === '*') {
      if (typeof current !== 'object' || Array.isArray(current))
        throw new Error('"*" requires an object');
      return Object.keys(current);
    } else {
      if (typeof current !== 'object' || Array.isArray(current))
        throw new Error(`Cannot get key "${part}" on ${typeof current}`);
      if (!(part in current))
        throw new Error(`Key "${part}" not found. Available: ${Object.keys(current).join(', ')}`);
      current = current[part];
    }
  }
  return current;
}

function applyTransform(value, transform) {
  switch (transform) {
    case 'keys':
      if (typeof value !== 'object' || value === null || Array.isArray(value))
        throw new Error("'keys' requires an object");
      return Object.keys(value).join('\n');
    case 'values':
      if (typeof value !== 'object' || value === null || Array.isArray(value))
        throw new Error("'values' requires an object");
      return Object.values(value).map(v => JSON.stringify(v)).join('\n');
    case 'length':
      if (typeof value === 'string' || Array.isArray(value)) return String(value.length);
      if (typeof value === 'object' && value !== null) return String(Object.keys(value).length);
      throw new Error("'length' requires a string, array, or object");
    case 'type':
      if (value === null) return 'null';
      if (Array.isArray(value)) return 'array';
      return typeof value;
    case 'pretty':
      return JSON.stringify(value, null, 2);
    case 'compact':
      return JSON.stringify(value);
    default:
      throw new Error(`Unknown transform "${transform}". Use: keys, values, length, type, pretty, compact`);
  }
}

function handleCall(params) {
  const jsonStr = params.json;
  if (typeof jsonStr !== 'string') return ['Error: "json" must be a string', true];

  let parsed;
  try { parsed = JSON.parse(jsonStr); }
  catch (e) { return [`Error: invalid JSON — ${e.message}`, true]; }

  const path = params.path ?? '.';
  let value;
  try { value = getPath(parsed, path); }
  catch (e) { return [`Error: ${e.message}`, true]; }

  const transform = params.transform;
  let result;
  if (transform) {
    try { result = applyTransform(value, transform); }
    catch (e) { return [`Error: ${e.message}`, true]; }
  } else {
    result = typeof value === 'string' ? value
           : value === null ? 'null'
           : typeof value === 'object' ? JSON.stringify(value, null, 2)
           : String(value);
  }

  return [result, false];
}

const rl = createInterface({ input: process.stdin, crlfDelay: Infinity });

rl.on('line', (line) => {
  const trimmed = line.trim();
  if (!trimmed) return;

  let msg;
  try { msg = JSON.parse(trimmed); }
  catch {
    process.stdout.write(JSON.stringify({ content: [{ type: 'text', text: 'JSON parse error' }], error: true }) + '\n');
    return;
  }

  let resp;
  if (msg.type === 'describe') {
    resp = DEFINITION;
  } else if (msg.type === 'call') {
    const [text, isError] = handleCall(msg.params ?? {});
    resp = { content: [{ type: 'text', text }], error: isError };
  } else {
    resp = { content: [{ type: 'text', text: `Unknown type: ${msg.type}` }], error: true };
  }

  process.stdout.write(JSON.stringify(resp) + '\n');
});

File: examples/tools/ruby/tool.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# template — Ruby plugin tool
#
# Renders simple Mustache-style templates: {{variable}} substitution,
# {{#if condition}} blocks, {{#each list}} loops. No gems required.
#
# Protocol:
#   in:  {"type":"describe"}
#   out: {"name":"template","description":"...","parameters":{...}}
#
#   in:  {"type":"call","call_id":"c1","params":{"template":"Hello, {{name}}!","vars":{"name":"World"}}}
#   out: {"content":[{"type":"text","text":"Hello, World!"}],"error":false}
#
# Usage:
#   ruby tool.rb
#   echo '{"type":"describe"}' | ruby tool.rb

require 'json'

# ── Template engine ──────────────────────────────────────────────────────────

class TemplateEngine
  def initialize(template, vars)
    @template = template
    @vars = deep_stringify(vars)
  end

  def render
    render_block(@template, @vars)
  end

  private

  def deep_stringify(obj)
    case obj
    when Hash  then obj.transform_keys(&:to_s).transform_values { |v| deep_stringify(v) }
    when Array then obj.map { |v| deep_stringify(v) }
    else            obj
    end
  end

  # Render a template string in the context of the given vars hash.
  def render_block(tmpl, vars)
    # Process block tags first (if/each), then simple substitutions.
    result = tmpl.dup

    # {{#each list}} ... {{/each}}
    result = result.gsub(/\{\{#each\s+(\w+)\}\}(.*?)\{\{\/each\}\}/m) do
      key, body = Regexp.last_match(1), Regexp.last_match(2)
      items = vars[key]
      unless items.is_a?(Array)
        raise "each: '#{key}' is not an array (got #{items.class})"
      end
      items.map.with_index do |item, idx|
        ctx = case item
              when Hash then vars.merge(item).merge('@index' => idx, '@first' => idx == 0, '@last' => idx == items.size - 1)
              else           vars.merge('@this' => item, '@index' => idx)
              end
        render_block(body, ctx)
      end.join
    end

    # {{#if condition}} ... {{else}} ... {{/if}}
    result = result.gsub(/\{\{#if\s+(\w+)\}\}(.*?)(?:\{\{else\}\}(.*?))?\{\{\/if\}\}/m) do
      key, then_body, else_body = Regexp.last_match(1), Regexp.last_match(2), Regexp.last_match(3)
      val = vars[key]
      truthy = val && val != false && val != '' && val != 0 && !(val.is_a?(Array) && val.empty?)
      render_block(truthy ? then_body : (else_body || ''), vars)
    end

    # {{#unless condition}} ... {{/unless}}
    result = result.gsub(/\{\{#unless\s+(\w+)\}\}(.*?)\{\{\/unless\}\}/m) do
      key, body = Regexp.last_match(1), Regexp.last_match(2)
      val = vars[key]
      truthy = val && val != false && val != '' && val != 0
      render_block(truthy ? '' : body, vars)
    end

    # {{{raw}}} — unescaped (triple braces, no HTML escaping needed here)
    result = result.gsub(/\{\{\{(\w+(?:\.\w+)*)\}\}\}/) do
      lookup(vars, Regexp.last_match(1).split('.')).to_s
    end

    # {{variable}} — standard substitution with dot-path support
    result = result.gsub(/\{\{(\w+(?:\.\w+)*)\}\}/) do
      lookup(vars, Regexp.last_match(1).split('.')).to_s
    end

    result
  end

  def lookup(vars, keys)
    current = vars
    keys.each do |k|
      return '' unless current.is_a?(Hash) && current.key?(k)
      current = current[k]
    end
    current
  end
end

# ── Tool definition ──────────────────────────────────────────────────────────

DEFINITION = {
  name: 'template',
  description:
    'Render a Mustache-style template with variable substitution, ' \
    'conditionals ({{#if}}/{{else}}/{{/if}}), loops ({{#each}}...{{/each}}), ' \
    'and negation ({{#unless}}). ' \
    'Useful for generating code, config files, reports, or formatted output from structured data.',
  parameters: {
    type: 'object',
    properties: {
      template: {
        type: 'string',
        description: 'Template string with {{variable}}, {{#if key}}, {{#each list}} tags'
      },
      vars: {
        type: 'object',
        description: 'Key-value pairs substituted into the template. Values can be strings, numbers, booleans, arrays, or nested objects.'
      }
    },
    required: %w[template vars]
  }
}.freeze

# ── Protocol loop ────────────────────────────────────────────────────────────

def ok_result(text)
  { content: [{ type: 'text', text: text }], error: false }
end

def err_result(text)
  { content: [{ type: 'text', text: text }], error: true }
end

def handle_call(params)
  tmpl = params['template']
  vars = params['vars'] || {}

  return err_result("Missing required parameter: 'template'") if tmpl.nil? || !tmpl.is_a?(String)
  return err_result("Parameter 'vars' must be an object") unless vars.is_a?(Hash)

  begin
    rendered = TemplateEngine.new(tmpl, vars).render
    ok_result(rendered)
  rescue => e
    err_result("Template error: #{e.message}")
  end
end

STDOUT.sync = true

STDIN.each_line do |line|
  line = line.strip
  next if line.empty?

  begin
    msg = JSON.parse(line)
  rescue JSON::ParserError => e
    puts JSON.generate(err_result("JSON parse error: #{e.message}"))
    next
  end

  response = case msg['type']
             when 'describe'
               DEFINITION
             when 'call'
               handle_call(msg['params'] || {})
             else
               err_result("Unknown message type: #{msg['type'].inspect}")
             end

  puts JSON.generate(response)
end
